local-volume/provisioner/pkg/deleter/blockcleaner.go

```diff
@@ -0,0 +1,247 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deleter
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/external-storage/local-volume/provisioner/pkg/common"
+	"k8s.io/api/core/v1"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	// ZeroDev is refers to zero device.
+	ZeroDev = "/dev/zero"
+	// DefaultBlockSize is the default block size to use in dd
+	DefaultBlockSize = "4096"
+	// ShredCommand the shred cli.
+	ShredCommand = "shred"
+	// MkfsCommand is the mkfs executable.
+	MkfsCommand = "mkfs"
+	// DDCommand is the dd executable.
+	DDCommand = "dd"
+	// WipeComand is the dd executable.
+	WipeComand = "wipefs"
+	// ShellCommand is the command to invoke shell
+	ShellCommand = "sh"
+
+	// DDFullErrExpected Error ouptput expected when running dd to fill a device
+	DDFullErrExpected = "No space left on device"
+	// NoErrExpected No error expected from running command
+	NoErrExpected = ""
+)
+
+// BlockCleaner Interface for invoking async script to do cleanup of block device.
+type BlockCleaner interface {
+	// DoCleanup deletes
+	DeleteBlockPV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error
+	isRunning(pvName string) bool
+	markRunning(pvName string)
+	markDone(pvName string)
+}
+
+var _ BlockCleaner = &BlockCleanerImpl{}
+
+// BlockCleanerImpl Implementation of BLockCleaner interface
+type BlockCleanerImpl struct {
+	*common.RuntimeConfig
+	mutex     sync.Mutex
+	procTable map[string]time.Time
+}
+
+// NewBlockCleaner returns a BlockCleaner
+func NewBlockCleaner(config *common.RuntimeConfig) *BlockCleanerImpl {
+	return &BlockCleanerImpl{RuntimeConfig: config,
+		procTable: make(map[string]time.Time),
+	}
+}
+
+func (v *BlockCleanerImpl) isRunning(pvName string) bool {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	_, ok := v.procTable[pvName]
+	return ok
+}
+
+func (v *BlockCleanerImpl) markRunning(pvName string) {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	v.procTable[pvName] = time.Now()
+}
+
+func (v *BlockCleanerImpl) markDone(pvName string) {
```

Comment:
Do we need to have a timeout for these operations? I don't think we can assume that things always complete successfully and keeping their state around forever. Its probably better to error out and drop tasks that aren't responsive over some time period.

Reply:
@ianchakeres I do want to track a process which is still deleting or writing to block as it is likely still active in some manner that makes the disk unusable until it terminates. The best way to actually drop the task is to kill it. Killing a process which is locked in I/O can quite often be a problem as it can ignore the signal when blocked in kernel I/O.  Also, with custom scripts, killing might have unintended consequences ( grand child processes stuck?). 

I think its highly preferable to let the issue be highlighted by a wedged PV and log messages. The admin fix would typically be to terminate the POD in which the provisioner is running, as all related processes would terminate. If we find that this happens often enough, then perhaps we could introduce a timeout which causes this self termination. I think its best to see some of these scenarios play out, and based on that design a timeout and termination approach.

Reply:
Even doing a kill -9 on a process stuck in the kernel won't help, and could potentially have bad consequences of causing any operation on that disk to hang.  The only way out I've found, is to reboot the node.  Maybe (as a separate feature), we should also have health probe on the provisioner if it ever gets into this situation, so we can notify the admin that some intervention may be needed.